### PR TITLE
tracing: add utxocache tracepoints

### PR DESCRIPTION
**Pull request by arnabsen1729**
This PR adds <count> tracepoints (including documentation).

If you are new to USDTs you are requested to go through [this](https://github.com/bitcoin/bitcoin/pull/22006) to get some context. 

It was proposed in https://github.com/bitcoin/bitcoin/issues/20981#issuecomment-802688809 some useful tracepoints that can be added. On the basis of that, this PR adds the following tracepoints:

<-- create a table -->

### `flush:coins_flush`
Traces every time caches and indexes are flushed.
- `int64` time (in microseconds)
- `uint32` flush mode
- `uint64` coins flushed 
- `uint64` coins memory usage

### `utxo:coins_added`
Traces every time a new coin is added to the cache
- `int32` height
- `int64` value
- `uint32` if the coin is coinbase
- `uint64` coins usage

### `utxo:coins_spent`
Traces every time a coin is spent
- `int32` height
- `int64` value
- `uint32` if the coin is coinbase
- `uint64` coins usage

### `utxo:coins_uncached`
Traces every time a UTXO with the given outpoint is removed from the cache
- `int32` height
- `int64` value
- `uint32` if the coin is coinbase
- `uint64` coins usage

### CODE REVIEW

File: contrib/tracing/log_utxo_coins.py [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+
+int trace_spend_coin(struct pt_regs *ctx) {
+  int idx = 0;
+  struct coin_data *data = data_arr.lookup(&idx);
+  if (data == NULL) return 1;
+  bpf_usdt_readarg(1, ctx, &data->height);
+  bpf_usdt_readarg(2, ctx, &data->value);
+  bpf_usdt_readarg(3, ctx, &data->isCoinBase);
+  bpf_usdt_readarg(4, ctx, &data->usage);
+  coins_added.perf_submit(ctx, data, sizeof(*data));
```
Comment (0xB10C) [inline contrib/tracing/log_utxo_coins.py]: should be `coins_spent.perf_submit(ctx, data, sizeof(*data));`

File: contrib/tracing/log_utxo_coins.py [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+
+int trace_uncached_coin(struct pt_regs *ctx) {
+    int idx = 0;
+    struct coin_data *data = data_arr.lookup(&idx);
+    if (data == NULL) return 1;
+    bpf_usdt_readarg(1, ctx, &data->height);
+    bpf_usdt_readarg(2, ctx, &data->value);
+    bpf_usdt_readarg(3, ctx, &data->isCoinBase);
+    bpf_usdt_readarg(4, ctx, &data->usage);
+    coins_added.perf_submit(ctx, data, sizeof(*data));
```
Comment (0xB10C) [inline contrib/tracing/log_utxo_coins.py]: should be `coins_uncached.perf_submit(ctx, data, sizeof(*data));`

File: doc/tracing.md [hunk 3]
```diff
@@ -129,6 +129,39 @@ Arguments passed:
...
 4. Size of the cache in bytes as `uint64`
 
+### Context `utxo`
+
+#### Tracepoint `utxo:coins_added`
+
+It is called *after* a new coin is added to the cache.
+
+Arguments passed:
+1. Height of the block the coin was added to as `int32`
```
Comment (0xB10C) [inline doc/tracing.md]: I'd find `1. Block height the coin was added to the UTXO-set as 'int32'` clearer.

File: doc/tracing.md [hunk 4]
```diff
@@ -129,6 +129,39 @@ Arguments passed:
 3. Number of coins flushed as `uint64`
 4. Size of the cache in bytes as `uint64`
 
+### Context `utxo`
+
+#### Tracepoint `utxo:coins_added`
+
+It is called *after* a new coin is added to the cache.
```
Comment (0xB10C) [inline doc/tracing.md]: Don't think it's important that the tracepoint is called _after_ the coin was added to the UTXO set cache.

Could be `Is called when a coin is added to the UTXO cache.`

File: doc/tracing.md [hunk 5]
```diff
@@ -129,6 +129,39 @@ Arguments passed:
...
+
+Arguments passed:
+1. Height of the block the coin was added to as `int32`
+2. Value of the coin as `int64`
+3. If the coin is a coinbase as `uint32`
+4. Cached coins usage as `uint64`
+
+#### Tracepoint `utxo:coins_spent`
+
+It is called *after* a coin is spent.
```
Comment (0xB10C) [inline doc/tracing.md]: Same, _after_ is not really relevant here.

File: src/coins.cpp [hunk 6]
```diff
@@ -229,8 +244,17 @@ bool CCoinsViewCache::Flush() {
 void CCoinsViewCache::Uncache(const COutPoint& hash)
 {
     CCoinsMap::iterator it = cacheCoins.find(hash);
+
```
Comment (0xB10C) [inline src/coins.cpp]: no whitespace changes please :)

File: contrib/tracing/log_utxo_coins.py [hunk 7]
```diff
@@ -0,0 +1,116 @@
+#!/usr/bin/env python3
+
+import sys
+from bcc import BPF, USDT
+from bcc.utils import printb
+
+""" Example script to log details about add/spend/uncache coin events.
+utilizing USDT probes and the utxo:coins_added, utxo:coins_spent,
+utxo:coins_uncached  tracepoint. """
```
Comment (0xB10C) [inline contrib/tracing/log_utxo_coins.py]: nit: space between `utxo:coins_uncached` and `tracepoint`.

File: contrib/tracing/log_utxo_coins.py [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+struct coin_data
+{
+  u32 height;
+  long value;
+  u32 isCoinBase;
+  u64 usage;
+};
+
+
+BPF_ARRAY(data_arr, struct coin_data, 1);
```
Comment (0xB10C) [inline contrib/tracing/log_utxo_coins.py]: I don't think you need a `BPF_ARRAY` here. This is only used when we can't fit the struct into the 512 byte eBPF stack. See [doc/tracing.md#ebpf-virtual-machine-limits](https://github.com/bitcoin/bitcoin/blob/master/doc/tracing.md#ebpf-virtual-machine-limits).

See e.g. [this](https://github.com/bitcoin/bitcoin/blob/92f3a4b4d0531976c9484f5b5075487e0925b619/contrib/tracing/p2p_monitor.py#L45-L54) from the P2P monitor as an example.

File: src/validation.cpp [hunk 9]
```diff
@@ -2145,6 +2145,7 @@ bool CChainState::FlushStateToDisk(
             if (!CoinsTip().Flush())
                 return AbortNode(state, "Failed to write to coin database");
             nLastFlush = nNow;
+            TRACE4(flush, coins_flushed, nNow.count(), mode, coins_count, coins_mem_usage);
```
Comment (0xB10C) [inline src/validation.cpp]: Could be multi-line, similar to the other tracepoints, too. And you likely want to cast here too.

File: doc/tracing.md [hunk 10]
```diff
@@ -115,6 +115,20 @@ details.
...
 
+### Context `flush`
+
+#### Tracepoint `flush:coins_flushed`
+
+Is called *after* the caches and indexes are flushed depending on the mode
+`CChainState::FlushStateToDisk` called with.
+
+Arguments passed:
+1. Time in micorseconds as `int64`
```
Comment (0xB10C) [inline doc/tracing.md]: I think our goal should be to log the flush duration and not the timestamp.

File: src/validation.cpp (hunk 9 above)
Comment (0xB10C) [reply src/validation.cpp]: Context shouldn't be `flush`. I think `utxocache` is clearer.

File: contrib/tracing/log_coins_flushed.py [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+
+struct data_t
+{
+  long ts;
+  u32 mode;
+  u64 coins;
+  u64 coins_mem_usage;
+};
+
+BPF_ARRAY(data_arr, struct data_t, 1);
```
Comment (0xB10C) [inline contrib/tracing/log_coins_flushed.py]: same as above. Doesn't need to be an BPF_ARRAY.

File: contrib/tracing/log_utxo_coins.py [hunk 12]
```diff
@@ -0,0 +1,116 @@
...
+utilizing USDT probes and the utxo:coins_added, utxo:coins_spent,
+utxo:coins_uncached  tracepoint. """
+
+program = """
+# include <uapi/linux/ptrace.h>
+struct coin_data
+{
+  u32 height;
+  long value;
+  u32 isCoinBase;
```
Comment (0xB10C) [inline contrib/tracing/log_utxo_coins.py]: are booleans supported? (I haven't checked)

**Review by 0xB10C — COMMENTED** [review]
first round of review

some general notes:
- [x] you can and should use `flake8` and `autopep8` to format Python scripts
- [x] I think we should pass information about the [outpoint](https://github.com/bitcoin/bitcoin/blob/92f3a4b4d0531976c9484f5b5075487e0925b619/src/primitives/transaction.h#L26-L30) (txid and output index) for UTXO set operations too. For passing the txid see arguments 1 and 7 of the `validation:block_connected` tracepoint.
- [x] I think naming the context `utxocache`, `utxosetcache`, or even `coinscache` is clearer. See https://github.com/bitcoin/bitcoin/blob/master/doc/tracing.md#adding-tracepoints-to-bitcoin-core
- [ ] ideally, we'd pass fCoinbase as `bool`, however I haven't checked if this is even possible

File: contrib/tracing/log_utxo_coins.py (hunk 1 above)
Comment (arnabsen1729) [reply contrib/tracing/log_utxo_coins.py]: Fixed :white_check_mark:

File: contrib/tracing/log_utxo_coins.py (hunk 2 above)
Comment (arnabsen1729) [reply contrib/tracing/log_utxo_coins.py]: Fixed :white_check_mark:

File: doc/tracing.md (hunk 3 above)
Comment (arnabsen1729) [reply doc/tracing.md]: Updated

File: doc/tracing.md (hunk 4 above)
Comment (arnabsen1729) [reply doc/tracing.md]: Okay, made the changes

File: src/validation.cpp (hunk 9 above)
Comment (arnabsen1729) [reply src/validation.cpp]: Made the refactoring, also added a new parameter of `outpoint.hash`

**Comment by 0xB10C** [comment]
> I think we should pass information about the outpoint (txid and output index) for UTXO set operations too. For passing the txid see arguments 1 and 7 of the validation:block_connected tracepoint.

An outpoint is a reference to a previous output. It consists of a txid (called `hash` in COutPoint) and the output index of the transaction (called `n` in COutPoint). See [`COutPoint`](https://github.com/bitcoin/bitcoin/blob/92f3a4b4d0531976c9484f5b5075487e0925b619/src/primitives/transaction.h#L25-L30).

I saw you added the `hash` aka txid of the outpoint as a C string. It would be good to pass the txid as bytes too. That could be done in the 7th argument (not accessible by bpftrace, and that is fine as bpftrace currently can't handle bytes arrays properly anyway) similar as we do in the `validation:block_connected` tracepoint. Additionally, we'd need the pass the output index (`n`) too. 

Would be good to log the outpoints in the bcc and bpftrace scripts too. Often outpoints are formatted as `<txid>:<index>`. And they should be documented.

**Comment by arnabsen1729** [comment]
@0xB10C I have used the `bool` datatype in the BCC script `log_coins_flushed.py` using ctype https://github.com/0xB10C/bitcoin/pull/1/commits/f43febb960beaa3be58269eaccebea14a783a9c4.

Also, updated the tracepoints context and names and added the `txid` as bytes to `add`, `spent` and `uncache` tracepoints.

File: contrib/tracing/log_coins_flushed.py [hunk 13]
```diff
@@ -1,12 +1,13 @@
 #!/usr/bin/env python3
 
 import sys
+import ctypes as ct
```
Comment (0xB10C) [inline contrib/tracing/log_coins_flushed.py]:
```suggestion
import ctypes
```

File: contrib/tracing/log_coins_flushed.py [hunk 14]
```diff
@@ -45,12 +49,26 @@
...
+class Data(ct.Structure):
+    _fields_ = [
+        ("ts", ct.c_long),
+        ("mode", ct.c_uint32),
+        ("coins", ct.c_uint64),
+        ("coins_mem_usage", ct.c_uint64),
+        ("is_flush_prune", ct.c_bool),
+        ("is_full_flush", ct.c_bool)
+    ]
+
```
Comment (0xB10C) [inline contrib/tracing/log_coins_flushed.py]:
```suggestion
class Data(ctypes.Structure):
    _fields_ = [
        ("ts", ctypes.c_long),
        ("mode", ctypes.c_uint32),
        ("coins", ctypes.c_uint64),
        ("coins_mem_usage", ctypes.c_uint64),
        ("is_flush_prune", ctypes.c_bool),
        ("is_full_flush", ctypes.c_bool)
    ]

```

I'd prefer not to shorten `ctypes` here. As it's 1) not so long and 2) is clearer when reviewing / reading the code.

File: contrib/tracing/log_coins_flushed.py [hunk 15]
```diff
@@ -45,12 +49,26 @@
     'ALWAYS'
 ]
 
+# define output data structure in Python
+class Data(ct.Structure):
+    _fields_ = [
+        ("ts", ct.c_long),
```
Comment (0xB10C) [inline contrib/tracing/log_coins_flushed.py]: Could this be a `c_uint64` too?
```suggestion
        ("ts", ct.c_uint64),
```

File: contrib/tracing/log_coins_flushed.py [hunk 16]
```diff
@@ -45,12 +49,26 @@
...
+        ("mode", ct.c_uint32),
+        ("coins", ct.c_uint64),
```
Comment (0xB10C) [inline contrib/tracing/log_coins_flushed.py]:
```suggestion
        ("coins_count", ct.c_uint64),
```

File: contrib/tracing/log_utxo_coins.py [hunk 17]
```diff
@@ -15,6 +15,7 @@
 struct coin_data
 {
   u32 height;
+  u32 index;
   char hash[MAX_HASH_SIZE];
   long value;
   u32 isCoinBase;
```
Comment (0xB10C) [inline contrib/tracing/log_utxo_coins.py]: should be `bool` too

File: src/validation.cpp [hunk 18]
```diff
@@ -2147,6 +2147,14 @@ bool CChainState::FlushStateToDisk(
...
             full_flush_completed = true;
         }
+        TRACE6(utxocache, flush,
+            (GetTimeMicros() - nNow.count()), // in microseconds (µs)
+            mode,
+            coins_count,
+            coins_mem_usage,
+            fFlushForPrune,
+            fDoFullFlush
+        );
```
Comment (0xB10C) [inline src/validation.cpp]: type castings? :)

File: src/coins.cpp [hunk 19]
```diff
@@ -229,8 +250,20 @@ bool CCoinsViewCache::Flush() {
...
     if (it != cacheCoins.end() && it->second.flags == 0) {
         cachedCoinsUsage -= it->second.coin.DynamicMemoryUsage();
+
+        TRACE7(utxocache, uncache,
+            hash.hash.ToString().c_str(),
+            (uint32_t)hash.n,
+            (uint32_t)it->second.coin.nHeight,
+            (int64_t)it->second.coin.out.nValue,
+            (uint)it->second.coin.fCoinBase,
```
Comment (0xB10C) [inline src/coins.cpp]:
```suggestion
            (bool)it->second.coin.fCoinBase,
```

File: contrib/tracing/log_coins_flushed.py [hunk 20]
```diff
@@ -0,0 +1,109 @@
...
+# USAGE:  ./contrib/tracing/log_flush.py path/to/bitcoind
+
+# BCC: The C program to be compiled to an eBPF program (by BCC) and loaded into
+# a sandboxed Linux kernel VM.
+program = """
+# include <uapi/linux/ptrace.h>
+
+struct data_t
+{
+  long ts;
```
Comment (0xB10C) [inline contrib/tracing/log_coins_flushed.py]:
```suggestion
  u64 duration;
```

File: contrib/tracing/log_coins_flushed.py [hunk 21]
```diff
@@ -0,0 +1,109 @@
...
+  u32 mode;
+  u64 coins;
```
Comment (0xB10C) [inline contrib/tracing/log_coins_flushed.py]:
```suggestion
  u64 coins_count;
```

File: contrib/tracing/log_coins_flushed.py [hunk 22]
```diff
@@ -0,0 +1,109 @@
...
+  bool is_flush_prune;
+  bool is_full_flush;
+};
+
+// BPF perf buffer to push the data to user space.
+BPF_PERF_OUTPUT(flush);
+
+int trace_flush(struct pt_regs *ctx) {
+  struct data_t data = {};
+
```
Comment (0xB10C) [inline contrib/tracing/log_coins_flushed.py]:
```suggestion
```

File: contrib/tracing/log_utxo_coins.py [hunk 23]
```diff
@@ -0,0 +1,116 @@
...
+        event = b["coins_uncached"].event(data)
+        print_event(event, "Uncache")
+
+    b["coins_added"].open_perf_buffer(handle_coins_added)
+    b["coins_spent"].open_perf_buffer(handle_coins_spent)
+    b["coins_uncached"].open_perf_buffer(handle_coins_uncached)
+
+    print("Logging Add, Spend and Uncache in the UTXO set. Ctrl-C to end...")
+    print("%-70s %-10s %-10s %-15s %-10s %s" %
+          ("txHash", "Action", "Height", "Satoshis", "Coin Base", "Usage"))
```
Comment (0xB10C) [inline contrib/tracing/log_utxo_coins.py]:
```suggestion
          ("Outpoint", "Action", "Height", "Value", "Coinbase", "Memory Usage"))
```

Comment (0xB10C) [reply contrib/tracing/log_utxo_coins.py]: Maybe also add cache size (count of cached utxos)?

File: contrib/tracing/log_utxo_coins.py [hunk 24]
```diff
@@ -0,0 +1,116 @@
...
+    bpf_usdt_readarg(5, ctx, &data.isCoinBase);
+    bpf_usdt_readarg(6, ctx, &data.usage);
+    coins_uncached.perf_submit(ctx, &data, sizeof(data));
+    return 0;
+}
+"""
+
+
+def print_event(event, action):
+    # print(event.hash.decode('ASCII'), action, event.height, event.value, "YES" if event.isCoinBase else "NO", event.usage)
```
Comment (0xB10C) [inline contrib/tracing/log_utxo_coins.py]:
```suggestion
```

**Review by 0xB10C — COMMENTED** [review]
Good work! I left a few more comments worth addressing. Also, some of the previous comments (e.g. https://github.com/0xB10C/bitcoin/pull/1/files#r692351276) are still open.

File: contrib/tracing/log_coins_flushed.py (hunk 15 above)
Comment (arnabsen1729) [reply contrib/tracing/log_coins_flushed.py]: Yup, there was a type `c_uint64` and it seemed to work without any issue

File: contrib/tracing/log_utxo_coins.py (hunk 17 above)
Comment (arnabsen1729) [reply contrib/tracing/log_utxo_coins.py]: Did you mean to `isCoinBase`?

Cause, `index` refers to the `COutPoint::n` here https://github.com/0xB10C/bitcoin/blob/7be143a960e2bb9ac81144f55c45731c1bb209c5/src/primitives/transaction.h#L26-L30

For the `isCoinBase` which refers to `Coin::fCoinBase` it is of `unsigned int` type.
https://github.com/0xB10C/bitcoin/blob/7be143a960e2bb9ac81144f55c45731c1bb209c5/src/coins.h#L30-L37

Comment (0xB10C) [reply contrib/tracing/log_utxo_coins.py]: yes, sorry if that was unclear. For me the comment is on the `isCoinbase` line.

image

Comment (0xB10C) [reply contrib/tracing/log_utxo_coins.py]: You could use Coin::IsCoinBase() which returns a bool

File: src/coins.cpp (hunk 19 above)
Comment (0xB10C) [reply src/coins.cpp]:
```suggestion
            (bool)it->second.coin.IsCoinBase(),
```

https://github.com/bitcoin/bitcoin/blob/b3a2b8c29fda5d924726d3087535bd34af7012b6/src/coins.h#L55-L57

File: src/coins.cpp [hunk 25]
```diff
@@ -95,6 +96,15 @@ void CCoinsViewCache::AddCoin(const COutPoint &outpoint, Coin&& coin, bool possi
     it->second.coin = std::move(coin);
     it->second.flags |= CCoinsCacheEntry::DIRTY | (fresh ? CCoinsCacheEntry::FRESH : 0);
     cachedCoinsUsage += it->second.coin.DynamicMemoryUsage();
+    TRACE7(utxocache, add,
+        outpoint.hash.ToString().c_str(),
+        (uint32_t)outpoint.n,
+        (uint32_t)coin.nHeight,
+        (int64_t)coin.out.nValue,
+        (uint)coin.fCoinBase,
```
Comment (0xB10C) [inline src/coins.cpp]: see https://github.com/0xB10C/bitcoin/pull/1/files?file-filters%5B%5D=.cpp#r700080149

File: src/coins.cpp [hunk 26]
```diff
@@ -120,6 +130,17 @@ bool CCoinsViewCache::SpendCoin(const COutPoint &outpoint, Coin* moveout) {
     CCoinsMap::iterator it = FetchCoin(outpoint);
     if (it == cacheCoins.end()) return false;
     cachedCoinsUsage -= it->second.coin.DynamicMemoryUsage();
+
+    TRACE7(utxocache, spent,
+        outpoint.hash.ToString().c_str(),
+        (uint32_t)outpoint.n,
+        (uint32_t)it->second.coin.nHeight,
+        (int64_t)it->second.coin.out.nValue,
+        (uint)it->second.coin.fCoinBase,
```
Comment (0xB10C) [inline src/coins.cpp]: https://github.com/0xB10C/bitcoin/pull/1/files?file-filters%5B%5D=.cpp#r700080149

File: contrib/tracing/log_utxo_coins.py (hunk 17 above)
Comment (0xB10C) [reply contrib/tracing/log_utxo_coins.py]: See https://github.com/0xB10C/bitcoin/pull/1/files?file-filters%5B%5D=.cpp#r700080149

File: contrib/tracing/log_utxo_coins.py (hunk 23 above)
Comment (0xB10C) [reply contrib/tracing/log_utxo_coins.py]: In src/coins.cpp you should be able to access cacheCoins in src/coins.h https://github.com/bitcoin/bitcoin/blob/7317e14a44c6efc545e6fb9bcedee7174e93a8fa/src/coins.h#L221

which is a `std::unordered_map` https://github.com/bitcoin/bitcoin/blob/7317e14a44c6efc545e6fb9bcedee7174e93a8fa/src/coins.h#L134

which implements size() https://en.cppreference.com/w/cpp/container/unordered_map/size

**Comment by 0xB10C** [comment]
closing this in favor of https://github.com/bitcoin/bitcoin/pull/22902 :)

**Pull request closed without merging**